### PR TITLE
Add -spirv-core-grammar option to load alternate spirv defs

### PR DESCRIPTION
**Pull request by expipiplus1**
Also embed a version to use by default

### CODE REVIEW

File: source/compiler-core/slang-spirv-core-grammar.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+);
+
+struct SPIRVSpec
+{
+    // List<UnownedStringSlice> copyright;
+    // UnownedStringSlice magic_number;
+    // UInt32 major_version;
+    // UInt32 minor_version;
+    // UInt32 revision;
+    List<InstructionPrintingClass> instruction_printing_class;
```
Comment (csyonghe) [inline source/compiler-core/slang-spirv-core-grammar.cpp:99]: any reason to use a different naming convention here?

File: source/slang/slang-diagnostic-defs.h [hunk 2]
```diff
@@ -245,6 +245,10 @@ DIAGNOSTIC(29100, Error, unrecognizedSPIRVOpcode, "unrecognized spirv opcode: $0
 DIAGNOSTIC(29101, Error, misplacedResultIdMarker, "the result-id marker must only be used in the last instruction of a spriv_asm expression")
 DIAGNOSTIC(29102, Note,  considerOpCopyObject, "consider adding an OpCopyObject instruction to the end of the spirv_asm expression")
 DIAGNOSTIC(29103, Note,  noSuchAddress, "unable to take the address of this address-of asm operand")
+DIAGNOSTIC(29104, Error, spirvInstructionWithoutResultId, "cannot use this 'x = $0...' syntax because $0 does not have a <result-id> operand")
+DIAGNOSTIC(29104, Error, spirvInstructionWithoutResultTypeId, "cannot use this 'x : <type> = $0...' syntax because $0 does not have a <result-type-id> operand")
+// This is a warning because we trust that people using the spirv_asm block know what they're doing
+DIAGNOSTIC(29104, Warning, spirvInstructionWithTooManyOperands, "Too many operands for $0 (expected max $1), did you forget a semicolon?")
```
Comment (csyonghe) [inline source/slang/slang-diagnostic-defs.h:251]: If we see a spvword that starts with `Op***`, we can tell it is a new inst, right?

File: source/slang/slang-emit-spirv.cpp [hunk 3]
```diff
@@ -3822,22 +3821,35 @@ struct SPIRVEmitContext
...
-                    opcode == SpvOpConstant ? getSection(SpvLogicalSectionID::ConstantsAndTypes)
-                    : opcode == SpvOpName ? getSection(SpvLogicalSectionID::DebugNames)
-                    : defaultParent;
+            const auto parentForOpCode = [this](SpvOp opcode, SpvInstParent* defaultParent) -> SpvInstParent*{
+                const auto info = m_grammarInfo->opInfos.lookup(opcode);
+                SLANG_ASSERT(info.has_value());
+                switch(info->class_)
+                {
+                    case SPIRVCoreGrammarInfo::OpInfo::TypeDeclaration:
+                    case SPIRVCoreGrammarInfo::OpInfo::ConstantCreation:
```
Comment (csyonghe) [inline source/slang/slang-emit-spirv.cpp:3830]: Can we add cases for `OpCapability` and `OpExtension` here?

File: source/slang/slang-parser.cpp [hunk 4]
```diff
@@ -6245,21 +6305,28 @@ namespace Slang
...
         {
             if(parser->LookAheadToken(TokenType::RBrace))
                 break;
             if(const auto inst = parseSPIRVAsmInst(parser))
                 asmExpr->insts.add(*inst);
             else
-                return nullptr;
+            {
+                failed = true;
+                // Recover to the semi or brace
```
Comment (csyonghe) [inline source/slang/slang-parser.cpp:6318]: EndOfFile

File: source/compiler-core/slang-spirv-core-grammar.cpp (hunk 1 above)
Comment (expipiplus1) [reply source/compiler-core/slang-spirv-core-grammar.cpp:99]: Yeah, it matches what's in the spirv json exactly, which allows for concise macros just below.

File: source/slang/slang-diagnostic-defs.h (hunk 2 above)
Comment (expipiplus1) [reply source/slang/slang-diagnostic-defs.h:251]: Sadly not, they can be used for example here: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSpecConstantOp

File: source/slang/slang-emit-spirv.cpp (hunk 3 above)
Comment (expipiplus1) [reply source/slang/slang-emit-spirv.cpp:3830]: yeah, good idea, I'll do that for this PR or the next one if this merges before I get to it

File: source/compiler-core/slang-spirv-core-grammar.cpp (hunk 1 above)
Comment (csyonghe) [reply source/compiler-core/slang-spirv-core-grammar.cpp:99]: Got it!

File: source/slang/slang-diagnostic-defs.h (hunk 2 above)
Comment (csyonghe) [reply source/slang/slang-diagnostic-defs.h:251]: Oh, that's too bad...

**Review by csyonghe — APPROVED** [review]

**Comment by csyonghe** [comment]
Closing, superseded by #3157.

**Pull request closed without merging**